mirage/server/dns_server_mirage.ml

```diff
@@ -76,20 +117,22 @@ module Make (P : Mirage_clock.PCLOCK) (M : Mirage_clock.MCLOCK) (TIME : Mirage_t
         Dns.read_tcp f >>= function
         | Error () -> drop ip ; Lwt.return_unit
         | Ok data ->
+          inc `Tcp_query;
           let now = Ptime.v (P.now_d_ps ()) in
           let elapsed = M.elapsed_ns () in
           let t, answer, notify, n = Dns_server.Primary.handle_buf !state now elapsed `Tcp ip port data in
           let n' = match n with
-            | Some `Keep -> tcp_out := Dns.IM.add ip flow !tcp_out ; None
+            | Some `Keep -> inc `Tcp_cache_add ; inc `Tcp_keep ; tcp_out := Dns.IM.add ip flow !tcp_out ; None
             | Some `Notify soa -> Some (`Notify soa)
             | Some `Signed_notify soa -> Some (`Signed_notify soa)
             | None -> None
           in
           maybe_update_state t >>= fun () ->
           maybe_notify recv_task t now elapsed n' >>= fun () ->
           (match answer with
-           | None -> Log.warn (fun m -> m "empty answer") ; Lwt.return_unit
+           | None -> Lwt.return_unit
```

Comment:
would it make sense to track number of empty answers too? If it made sense at some point as a debug message for tcp+udp, maybe a metric would be good to track down errors in the future?

Reply:
that is a good suggestion, although this is better done in the core dns library: here we have no information about the contents of a packet -- i.e. short packets (<12 byte), or answers (bit 17 = 1) would and should not provoke any replies -- and thus would be wrongly counted). I removed the log message since it was misleading.